apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.SingleChannel.cs

```diff
@@ -1109,61 +1109,121 @@ private void UpdateReceiptModAckStatus(IEnumerable<string> ids, bool? status)
+            private const int MaxAuthExceptionsBeforeFailing = 4;
+
+            private const int ExceptionLimit = 100;
+
             private readonly IClock _clock;
-            // Only used for timing.
-            private readonly RetrySettings _retrySettings;
+            private readonly ILogger _logger;
+            private readonly RetrySettings _backoffTiming;
             private readonly TimeSpan _disconnectBackoff;
 
+            private DateTime _currentStartTimestamp;
             private readonly List<RpcException> _exceptions;
             private DateTime? _firstExceptionTimestamp;
-            public TimeSpan? Backoff { get; private set; }
+            internal TimeSpan? Backoff { get; private set; }
 
-            public RetryState(IClock clock, RetrySettings retrySettings, TimeSpan disconnectBackoff)
+            internal RetryState(IClock clock, ILogger logger, RetrySettings backoffTiming, TimeSpan disconnectBackoff)
             {
                 _clock = clock;
-                _exceptions = new List<RpcException>();
-                _firstExceptionTimestamp = null;
-                _retrySettings = retrySettings;
+                _logger = logger;
+                _backoffTiming = backoffTiming;
                 _disconnectBackoff = disconnectBackoff;
+                _exceptions = new();
+                _firstExceptionTimestamp = null;
                 Backoff = null;
             }
 
             /// <summary>
-            /// Checks whether the given exception should be retried, without updating any state.
-            /// </summary>
-            public bool ShouldRetry(Exception exception) =>
-                exception?.As<RpcException>()?.IsRecoverable() ?? false;
-
-            /// <summary>
-            /// Records the given exception in the retry state, updating the backoff accordingly.
+            /// Checks whether the given exception should be retried, updating the state including the next backoff.
             /// </summary>
-            public void OnFailure(Exception exception)
+            internal bool RecordFailureAndCheckForRetry(Exception exception)
             {
-                _firstExceptionTimestamp ??= _clock.GetCurrentDateTimeUtc();
-                if (exception?.As<RpcException>() is RpcException rpcException && _exceptions.Count < 100)
+                var now = _clock.GetCurrentDateTimeUtc();
+                _firstExceptionTimestamp ??= now;
+                // This won't be used if we decide not to retry anyway.
+                Backoff = _backoffTiming.NextBackoff(Backoff ?? TimeSpan.Zero);
+
+                // Don't retry any non-RpcExceptions
+                if (exception.As<RpcException>() is not RpcException rpcEx)
+                {
+                    return false;
+                }
+                if (_exceptions.Count < ExceptionLimit)
```

Comment:
I don't see `_exceptions.Count` being used anywhere else? I'm probably misising something but it seems like after the limit, we continue retrying as usual except that we don't store the exceptions?

Reply:
Discussed offline: while that *was* the previous intended behavior (just to prevent a memory leak), we're now going to fail (no retry) after this many exceptions.